code/ui/blocks/src/blocks/Source.tsx

```diff
@@ -94,50 +99,52 @@ export const useSourceProps = (
   docsContext: DocsContextProps<any>,
   sourceContext: SourceContextProps
 ): PureSourceProps & SourceStateProps => {
-  const { id: primaryId, parameters } = docsContext.storyById();
-
-  const codeProps = props as CodeProps;
-  const singleProps = props as SingleSourceProps;
-  const multiProps = props as MultiSourceProps;
-
-  let { format, code: source } = codeProps; // prefer user-specified code
-
-  const targetIds = multiProps.ids || [singleProps.id || primaryId];
-  const storyIds = targetIds.map((targetId) => {
-    return targetId;
-  });
-
-  const stories = useStories(storyIds, docsContext);
-  if (!stories.every(Boolean)) {
+  const storyIds = props.ids || (props.id ? [props.id] : []);
+  const storiesFromIds = useStories(storyIds, docsContext);
+  if (!storiesFromIds.every(Boolean)) {
     return { error: SourceError.SOURCE_UNAVAILABLE, state: SourceState.NONE };
   }
 
-  if (!source) {
-    // just take the format from the first story, given how they're all concatinated together...
-    // TODO: we should consider sending an event with all the sources separately, instead of concatenating them here
-    ({ format } = getStorySource(storyIds[0], sourceContext));
-    source = storyIds
-      .map((storyId, idx) => {
-        const { code: storySource } = getStorySource(storyId, sourceContext);
-        const storyObj = stories[idx] as PreparedStory;
-        return getSnippet(storySource, storyObj);
+  // The check didn't actually change the type.
+  let stories: PreparedStory[] = storiesFromIds as PreparedStory[];
+  if (props.of) {
+    const resolved = docsContext.resolveModuleExport(props.of);
+    if (resolved.type !== 'story')
+      throw new Error(`Invalid value passed to the 'of' prop, it should be a story export.`);
```

Comment:
could be `useOf` here as it supports throwing on invalid types. however you'd have to move the call all the way up to the beginning of the component.

Reply:
I decided against it because:

1. I already had the context, so I didn't need a hook, which like you said, would require some gymnastics to make sure I obeyed the rules of hooks.

2. The throwing on invalid types actually doesn't help that much because TS wants me to check it inside this function anyway. We could possibly do some TS smarts to make the `useOf` function return something typed based on the `validTypes` argument, but I couldn't immediately see how.

Reply:
OK, I'm going to merge this PR and make changes we discussed in a follow up.